client/monitor_endpoint.go

```diff
@@ -0,0 +1,75 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"io"
+	"time"
+
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/command/agent/monitor"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/ugorji/go/codec"
+
+	metrics "github.com/armon/go-metrics"
+	log "github.com/hashicorp/go-hclog"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+)
+
+type Monitor struct {
+	c *Client
+}
+
+func NewMonitorEndpoint(c *Client) *Monitor {
+	m := &Monitor{c: c}
+	m.c.streamingRpcs.Register("Agent.Monitor", m.monitor)
+	return m
+}
+
+func (m *Monitor) monitor(conn io.ReadWriteCloser) {
+	defer metrics.MeasureSince([]string{"client", "monitor", "monitor"}, time.Now())
+	defer conn.Close()
+
+	// Decode arguments
+	var req cstructs.MonitorRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&req); err != nil {
+		handleStreamResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// Check acl
+	if aclObj, err := m.c.ResolveToken(req.QueryOptions.AuthToken); err != nil {
+		handleStreamResultError(err, helper.Int64ToPtr(403), encoder)
+		return
+	} else if aclObj != nil && !aclObj.AllowNsOp(req.Namespace, acl.NamespaceCapabilityReadFS) {
```

Comment:
need to verify this is the right acl check

Reply:
You probably want to require `node:read` or `node:write` here. (Although it's a read only operation, logs could potentially be sensitive, and so write may be more appropriate. Given the work we're doing to improve observability we may actually want to consider adding a cluster debug permission for this + pprof endpoints cc @schmichael)

Reply:
Since this is supported by servers as well, I think we should go with `agent:read`. It also matches Consul's behavior: https://www.consul.io/api/agent.html#stream-logs

It's tempting to do something like `agent:write` as a hedge against accidently leaking credentials (eg Vault tokens), but I think logging Vault tokens would be a serious security issue even without this API.

As a followup to this let's do a big audit of log lines to ensure no secrets are logged.